channels/disp/server/disp_main.c

```diff
@@ -0,0 +1,569 @@
+	{
+		WLog_ERR(TAG, "MonitorLayoutSize is set to %"PRIu32". expected %"PRIu32"", pdu.MonitorLayoutSize,
+		         DISPLAY_CONTROL_MONITOR_LAYOUT_SIZE);
+		return ERROR_INVALID_DATA;
+	}
+
+	Stream_Read_UINT32(s, pdu.NumMonitors); /* NumMonitors (4 bytes) */
+
+	if (pdu.NumMonitors > context->MaxNumMonitors)
+	{
+		WLog_ERR(TAG, "NumMonitors (%"PRIu32")> server MaxNumMonitors (%"PRIu32")", pdu.NumMonitors,
+		         context->MaxNumMonitors);
+		return ERROR_INVALID_DATA;
+	}
+
+	if (Stream_GetRemainingLength(s) < DISPLAY_CONTROL_MONITOR_LAYOUT_SIZE * pdu.NumMonitors)
+	{
+		WLog_ERR(TAG, "not enough data!");
+		return ERROR_INVALID_DATA;
+	}
+
+	pdu.Monitors = (DISPLAY_CONTROL_MONITOR_LAYOUT*) calloc(pdu.NumMonitors,
+	               sizeof(DISPLAY_CONTROL_MONITOR_LAYOUT));
+
+	if (!pdu.Monitors)
+	{
+		WLog_ERR(TAG, "disp_recv_display_control_monitor_layout_pdu(): calloc failed!");
+		return CHANNEL_RC_NO_MEMORY;
+	}
+
+	WLog_DBG(TAG, "disp_recv_display_control_monitor_layout_pdu: NumMonitors=%"PRIu32"",
+	         pdu.NumMonitors);
+
+	for (index = 0; index < pdu.NumMonitors; index++)
+	{
+		monitor = &(pdu.Monitors[index]);
+		Stream_Read_UINT32(s, monitor->Flags); 				/* Flags (4 bytes) */
+		Stream_Read_UINT32(s, monitor->Left); 				/* Left (4 bytes) */
+		Stream_Read_UINT32(s, monitor->Top); 				/* Top (4 bytes) */
+		Stream_Read_UINT32(s, monitor->Width); 				/* Width (4 bytes) */
+		Stream_Read_UINT32(s, monitor->Height); 			/* Height (4 bytes) */
+		Stream_Read_UINT32(s, monitor->PhysicalWidth); 		/* PhysicalWidth (4 bytes) */
+		Stream_Read_UINT32(s, monitor->PhysicalHeight); 	/* PhysicalHeight (4 bytes) */
+		Stream_Read_UINT32(s, monitor->Orientation); 		/* Orientation (4 bytes) */
+		Stream_Read_UINT32(s, monitor->DesktopScaleFactor); /* DesktopScaleFactor (4 bytes) */
+		Stream_Read_UINT32(s, monitor->DeviceScaleFactor); 	/* DeviceScaleFactor (4 bytes) */
+		WLog_DBG(TAG,
+		         "\t%d : Flags: 0x%08"PRIX32" Left/Top: (%"PRId32",%"PRId32") W/H=%"PRIu32"x%"PRIu32")", index,
+		         monitor->Flags, monitor->Left, monitor->Top,
+		         monitor->Width,
+		         monitor->Height);
+		WLog_DBG(TAG, "\t   PhysicalWidth: %"PRIu32" PhysicalHeight: %"PRIu32" Orientation: %"PRIu32"",
+		         monitor->PhysicalWidth, monitor->PhysicalHeight,
+		         monitor->Orientation);
+
+		if (!disp_server_is_monitor_layout_valid(monitor))
+		{
+			error = ERROR_INVALID_DATA;
+			goto out;
+		}
+	}
+
+	if (context)
+		IFCALLRET(context->DispMonitorLayout, error, context, &pdu);
+
+out:
+	free(pdu.Monitors);
+	return error;
+}
+
+
+static UINT disp_server_receive_pdu(DispServerContext* context, wStream* s)
+{
+	UINT error = CHANNEL_RC_OK;
+	size_t beg, end;
+	DISPLAY_CONTROL_HEADER header;
+	beg = Stream_GetPosition(s);
+
+	if ((error = disp_read_header(s, &header)))
+	{
+		WLog_ERR(TAG, "rdpgfx_read_header failed with error %"PRIu32"!", error);
+		return error;
+	}
+
+	switch (header.type)
```

Comment:
Here it could be nice to create a static sub stream pointing on the packet (pointing at `Stream_Pointer(s)` with length `header.length`):
Something like:
```
wStream packetStream;
...
Stream_StaticInit(&packetStream, Stream_Pointer(s), header.length);

...

if ((error = disp_recv_display_control_monitor_layout_pdu(&packetStream, context)))
...
```

Then you can remove the check with `beg` and `end` which can be erroneous if the announced `header.length` is bigger than the effective payload.

Reply:
Not sure I want this, because after the call to disp_recv_display_control_monitor_layout_pdu or any other PDU (if there were more), you need to adjust the position of the original stream `s`:

```c
static UINT disp_server_receive_pdu(DispServerContext* context, wStream* s)
{
	UINT error = CHANNEL_RC_OK;
	wStream packetStream;
	DISPLAY_CONTROL_HEADER header;
	if ((error = disp_read_header(s, &header)))
	{
		WLog_ERR(TAG, "rdpgfx_read_header failed with error %"PRIu32"!", error);
		return error;
	}

	Stream_StaticInit(&packetStream, Stream_Pointer(s), header.length);

	switch (header.type)
	{
		case DISPLAY_CONTROL_PDU_TYPE_MONITOR_LAYOUT:
			if ((error = disp_recv_display_control_monitor_layout_pdu(&packetStream, context)))
				WLog_ERR(TAG, "disp_recv_display_control_monitor_layout_pdu "
				         "failed with error %"PRIu32"!", error);

			break;

		default:
			error = CHANNEL_RC_BAD_PROC;
			WLog_WARN(TAG, "Received unknown PDU type: %"PRIu32"", header.type);
			break;
	}

	Stream_Seek(s, Stream_GetPosition(s) + header.length);
	return error;
}
```
what do you both think? @akallabeth @hardening

Reply:
Looks fine except that it's `Stream_Seek(s, header.length)`

Reply:
@hardening don´t get your reasoning here, why create a copy of the stream (not content wise but meta info)?

Reply:
@akallabeth I agree

Reply:
@kubistika @akallabeth reopening that conversation. `Stream_StaticInit` doesn't create a copy, it just create a "substream" pointing on the same data. Without a substream, `disp_recv_display_control_monitor_layout_pdu`could read more than `header.length`, I agree that it is very unlikely to happen (as we have only 1 message at a time).

General topic: we have a lot of places where we should replace that beg/end computations with `Stream_StaticInit` and work on the substream, to prevent these "reads out of the pdu".

Reply:
Sorry to bother, but I'd really like to have these `Stream_StaticInit` changes.